cxx4opencl/address_spaces.txt

```diff
@@ -0,0 +1,235 @@
+// Copyright 2019 The Khronos Group. This work is licensed under a
+// Creative Commons Attribution 4.0 International License; see
+// http://creativecommons.org/licenses/by/4.0/
+
+[[opencl_c_variadic_macro]]
+=== Address spaces
+
+{cpp} for OpenCL inherits address space behavior from `OpenCL C v2.0 s6.5`.
+
+This section only documents behavior related to {cpp} features. For example
+conversion rules are extended from the qualification conversion but the
+compatibility is determined using notation of sets and overlapping of address
+spaces from Embedded C (`ISO/IEC JTC1 SC22 WG14 N1021 s3.1.3`). For OpenCL
+it means that implicit conversions are allowed from a named address space
+(except for `__constant`) to generic (`OpenCL C v2.0 6.5.5`). Reverse
+conversion is only allowed explicitly. The `__constant` address space does
+not overlap with any other and therefore no valid conversion between
+`__constant` and other address spaces exists. Most of the rules follow this
+logic.
+
+==== Casts
+
+C-style casts follow `OpenCL C v2.0 rules (s6.5.5)`. All cast operators
+permit conversion to generic address space implicitly. However converting
+from generic to named address spaces can only be done using
+`addrspace_cast`. Note that conversions between `__constant` and any other
+other address space are disallowed.
+
+TODO: Example with `addrspace_cast`.
+
+[[addrspace-deduction]]
+==== Deduction
+
+Address spaces are not deduced for:
+
+  * non-pointer/non-reference template parameters or any dependent types
+    except for template specializations.
+  * non-pointer/non-reference class members except for static data members
+    that are deduced to `__global` address space.
+  * non-pointer/non-reference alias declarations.
+  * decltype expressions.
+
+[source,c]
+----------
+template <typename T>
+void foo() {
+  T m; // address space of m will be known at template instantiation time.
+  T * ptr; // ptr points to generic address space object.
+  T & ref = ...; // ref references an object in generic address space.
+};
+
+template <int N>
+struct S {
+  int i; // i has no address space
```

Comment:
Usage or lack of `.` at the end of the comment is inconsistent with the rest of the document.